### ISSUE TITLE
[PaaS] Start MongoDB on Openshift

### ISSUE THREAD

**Issue by hekonsek** [rhiot/rhiot#581]
I'd like to start MongoDb with persistent volume on OpenShift. I'd like Mongo to be accessible using `MONGODB` service name.

The commands I have figured out so far:

```
docker exec -it origin oc new-app mongo
docker exec -it origin oc expose dc mongo --generator=service/v1 --name=mongodb
docker exec -it origin oc delete svc mongo
docker exec -it origin oc volume dc/mongo --remove --confirm
docker exec -it origin oc volume dc/mongo --add --name=mongodb-data --mount-path=/data/db
```

Now the question is - how can I expose MongoDB outside the Kubernetes?

```
docker exec -it origin oc get service                                                              
NAME         CLUSTER-IP       EXTERNAL-IP   PORT(S)                 AGE
kubernetes   172.30.0.1       <none>        443/TCP,53/UDP,53/TCP   4d
mongodb      172.30.159.134   <none>        27017/TCP               33m
```

How can I assign external IP to the mongodb?

**Comment by hekonsek** [comment]
@rimolive Any idea how to assign external IP address to MongoDB? :)

**Comment by rimolive** [comment]
@hekonsek, you must create a route in OpenShift with SNI passthrough. This is a link for the routes I created for A-MQ to expose all 4 protocols: https://github.com/jboss-openshift/ce-testsuite/blob/master/amq/src/test/resources/amq-routes.json

As you may know, using SNI implies using SSL/TLS protocol.

Hope that helps.

**Comment by hekonsek** [comment]
For the beginning of our migration to OS3, I propose to start MongoDB via the OS3 instance started in our PaaS script [1] (now we run MongoDB as plain Docker container). That MongoDB instance should have port 27017 exposed (temporarily), so services not migrated to OS3 should be able to access MongoDB via `localhost:27017`.

Can you create pull request with such change? That would be awesome for the beginning and we could continue the migration.

[1] https://github.com/rhiot/rhiot/blob/master/cloudplatform/install/rhiot-cloud-platform#L167

**Comment by rimolive** [comment]
Based on your proposal for rhiot-cloud-platform, then we should consider check if oc command is in PATH env var. This command will be very useful because it will create the OpenShift project, add the mongodb-persistent template to it and port-forward the MongoDB ports to localhost (all I'm suggesting is based on how the script is written so far, so let me know if that's what you plan).

**Comment by hekonsek** [comment]
I propose to use dockerized OC, like `docker exec -it openshift-server oc get service`. We assume that Docker is installed anyway :) .

**Comment by hekonsek** [comment]
So I would say that we should replace existing MongoDB setup with the following lines: 

```
docker exec -it origin oc new-app mongo
docker exec -it origin oc expose dc mongo --generator=service/v1 --name=mongodb
docker exec -it origin oc delete svc mongo
docker exec -it origin oc volume dc/mongo --remove --confirm
docker exec -it origin oc volume dc/mongo --add --name=mongodb-data --mount-path=/data/db
```

Plus, we should somehow expose MongoDB to `localhost` (via route I guess).

What do you think?

**Comment by rimolive** [comment]
Yeah, that's exactly what I was thinking. But there are some things I'm doing:

* I'm using the OpenShift 3 mongodb-persistent template to provision MongoDB
* The `oc volume` command you are using seems missing something, do I did it different: `oc volume dc/mongodb --add --type=emptyDir  --name=mongodb-data --overwrite`. Let me know if I can use it for now

To expose the MongoDB to localhost, just run `oc port-forward` command. The only drawback is that you must have the pod name, which makes a bit difficult to expose via script. But I'll see what I can do. =D

**Comment by hekonsek** [comment]
Sounds cool! It's good to have somebody knowing OS better than myself :) .

Would you be so kind and shoot a pool request with an appropriate changes?

**Comment by rimolive** [comment]
I sent https://github.com/rhiot/rhiot/pull/589, let me know if that helps.

**Comment by hekonsek** [comment]
Command for exposing port would be:

    docker exec -ti openshift-server oc port-forward $(docker exec -it openshift-server oc get pod | awk 'NR>1' | awk '{print $1}' | tail -n 1) 27017

**Comment by rimolive** [comment]
Doh! Yup, my fault. =/

**Comment by hekonsek** [comment]
No worries :)

Now I have:

```
EXEC : docker exec -ti openshift-server oc port-forward 27017
cannot enable tty mode on non tty input
```

Any idea how to handle this?

**Comment by rimolive** [comment]
I'll see if I can find a solution for this.

**Comment by hekonsek** [comment]
Many thanks in advance. BTW Can we use routes to expose this port?

**Comment by rimolive** [comment]
Yes, but it requires TLS passthrough to do that. That means: for MongoDB port 27017 is exposed, but outside OpenShift the port is 443.

**Comment by hekonsek** [comment]
Hmmmm, no other options for exposing ports? Only port-forward and TLS
passthrough?

śr., 11.05.2016 o 23:26 użytkownik Ricardo Martinelli de Oliveira <
notifications@github.com> napisał:

> Yes, but it requires TLS passthrough to do that. That means: for MongoDB
> port 27017 is exposed, but outside OpenShift the port is 443.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rhiot/rhiot/issues/581#issuecomment-218595300>
>
-- 
Henryk Konsek
https://linkedin.com/in/hekonsek

**Comment by hekonsek** [comment]
BTW The latest version of the script could look like:

```
log_info "MongoDB"
sleep 10
if [ $(docker exec -it openshift-server oc get service | grep mongo | wc -l) -eq 0 ]; then
    log_info "MongoDB doesn't exist. Creating..."
    exe docker exec -it openshift-server oc new-app mongo
    exe docker exec -it openshift-server oc expose dc mongo --generator=service/v1 --name=mongodb
    exe docker exec -it openshift-server oc delete svc mongo
fi
sleep 20
docker exec -ti openshift-server oc port-forward $(docker exec -it openshift-server oc get pod | awk 'NR>1' | awk '{print $1}' | tail -n 1) 27017 &
log_info "MongoDB started"
```

**Comment by rimolive** [comment]
Unfortunately, yes. =/

The script looks good. I'm still working on the port-forward issue
